lib/opam_solve.ml

```diff
@@ -251,6 +254,65 @@ module Make_solver (Context : OPAM_MONOREPO_CONTEXT) :
       rolemap []
     |> List.filter_map ~f:Solver.package_name
 
+  let find_version_restriction component =
+    let open Option.O in
+    let notes = Solver.Diagnostics.Component.notes component in
+    let* restriction =
+      List.find_map
```

Comment:
I'm not sure that's how we should do it here. I think we should rather merge the different restrictions we find together. For instance the solver could fail to find a solution because some package depend on `a > 1.2` and another on `a < 2.0`.
There could be versions that match either of those, for instance `a.1.2` and `a.2.0` but none that match both.

Reply:
Instead of `find_map` I could `filter_map` and `&&` all the constraints together (and potentially deduplicate them before, since I assume `a > 1.2` and `a < 2.0` would appear multiple times). Does that sound sensible as a solution to you?

This would be a good case of writing a test case first and implementing the solution in a second step.

Reply:
Yeah that sounds great!

I expect proper deduplication might require a bit of work to get right. For instance, `a < 2.2` would be redundant with `a < 2.0`. In such a case we'd prefer to only display the latter. I think it's fine if we iteratively improve this, I think users would prefer a correct but redundant output to an incorrect but concise one!

Reply:
Nah I think it's fine, we can refine it later, this is already helpful in its current state!